connect/mirror/src/main/java/org/apache/kafka/connect/mirror/OffsetSyncStore.java

```diff
@@ -25,17 +25,39 @@
 import org.apache.kafka.common.utils.Utils;
 import org.apache.kafka.connect.util.KafkaBasedLog;
 import org.apache.kafka.connect.util.TopicAdmin;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
+import java.util.Arrays;
 import java.util.Collections;
 import java.util.Map;
 import java.util.Optional;
 import java.util.OptionalLong;
 import java.util.concurrent.ConcurrentHashMap;
 
-/** Used internally by MirrorMaker. Stores offset syncs and performs offset translation. */
+/**
+ * Used internally by MirrorMaker. Stores offset syncs and performs offset translation.
+ * <p>A limited number of offset syncs can be stored per TopicPartition, in a way which provides better translation
+ * later in the topic, closer to the live end of the topic.
+ * This maintains the following invariants for each topic-partition in the in-memory sync storage:
+ * <ul>
+ *     <li>Invariant A: syncs[0] is the latest offset sync from the syncs topic</li>
+ *     <li>Invariant B: For each i,j, i <= j: syncs[j].upstream <= syncs[i].upstream < syncs[j].upstream + 2^j</li>
```

Comment:
Is it possible (and worthwhile) to establish a third invariant about a lower bound for `syncs[j].upstream`? Otherwise, what's laid out here could be satisfied by always populating the offset sync array with the latest sync, which wouldn't be terribly useful for addressing the ticket.

Reply:
I split this invariant into two parts: the upper bound and lower bound.
I also made both of the bounds stronger to promote a more effective distribution of syncs.

This means that the old logic of bit comparisons is gone now, and the invariants actually drive the algorithm directly, instead of the bit comparisons happening to satisfy the invariants but not actually having anything to do with them.